File: javascript/defaults/application/routing.js
//
// Prudence Application Routing
//

importClass(
	java.lang.ClassLoader,
	java.io.File,
	java.util.ArrayList,
	java.util.concurrent.ConcurrentHashMap,
	org.restlet.routing.Router,
	org.restlet.routing.Redirector,
	org.restlet.routing.Template,
	org.restlet.resource.Finder,
	org.restlet.resource.Directory,
	org.restlet.engine.application.Encoder,
	com.threecrickets.scripturian.util.DefrostTask,
	com.threecrickets.scripturian.document.DocumentFileSource,
	com.threecrickets.prudence.PrudenceRouter,
	com.threecrickets.prudence.util.PreheatTask,
	com.threecrickets.prudence.util.PhpExecutionController)

var classLoader = ClassLoader.systemClassLoader

//
// Utilities
//

// Makes sure we have slashes where we expect them
function fixURL(url) {
	url = url.replace(/\/\//g, '/') // no doubles
	if(url.length > 0 && url[0] == '/') { // never at the beginning
		url = url.slice(1)
	}
	if(url.length > 0 && url[url.length - 1] != '/') { // always at the end
		url += '/'
	}
	return url
}

//
// Internal router
//

component.internalRouter.attach('/' + applicationInternalName + '/', applicationInstance).matchingMode = Template.MODE_STARTS_WITH

//
// Hosts
//
// Note that the application's context will not be created until we attach the application to at least one
// virtual host. See defaults/instance/hosts.js for more information.
//

var addTrailingSlash = new Redirector(applicationInstance.context, '{ri}/', Redirector.MODE_CLIENT_PERMANENT)

print(applicationInstance.name + ': ')
for(var i in hosts) {
	var entry = hosts[i]
	var host = entry[0]
	var url = entry[1]
	if(!url) {
		url = applicationDefaultURL
	}
	print('"' + url + '" on ' + host.name)
	host.attach(url, applicationInstance).matchingMode = Template.MODE_STARTS_WITH
	if(url != '/') {
		if(url[url.length - 1] == '/') {
			url = url.slice(0, -1)
		}
		host.attach(url, addTrailingSlash).matchingMode = Template.MODE_EQUALS
	}
	if(i < hosts.length - 1) {
		print(', ')
	}
}
print('.\n')

var applicationGlobals = applicationInstance.context.attributes

applicationGlobals.put('com.threecrickets.prudence.component', component)
var cache =  component.context.attributes.get('com.threecrickets.prudence.cache')
if(cache) {
	applicationGlobals.put('com.threecrickets.prudence.cache', cache)
}

//
// Inbound root
//

var router = new PrudenceRouter(applicationInstance.context)
router.routingMode = Router.MODE_BEST_MATCH
applicationInstance.inboundRoot = router

//
// Add trailing slashes
//

for(var i in urlAddTrailingSlash) {
	urlAddTrailingSlash[i] = fixURL(urlAddTrailingSlash[i])
	if(urlAddTrailingSlash[i].length > 0) {
		if(urlAddTrailingSlash[i][urlAddTrailingSlash[i].length - 1] == '/') {
			// Remove trailing slash for pattern
			urlAddTrailingSlash[i] = urlAddTrailingSlash[i].slice(0, -1)
		}
		router.attach(urlAddTrailingSlash[i], addTrailingSlash)
	}
}

var languageManager = executable.manager

//
// Handlers
//

var handlersDocumentSource = new DocumentFileSource(applicationBase + handlersBasePath, applicationBasePath + handlersBasePath, handlersDefaultName, 'js', handlersMinimumTimeBetweenValidityChecks)
router.filterDocumentSource = handlersDocumentSource
router.filterLanguageManager = languageManager

//
// Dynamic web
//

var dynamicWebDocumentSource = new DocumentFileSource(applicationBase + dynamicWebBasePath, applicationBasePath + dynamicWebBasePath, dynamicWebDefaultDocument, 'js', dynamicWebMinimumTimeBetweenValidityChecks)
var cacheKeyPatternHandlers = new ConcurrentHashMap()
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.languageManager', languageManager)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.defaultLanguageTag', 'javascript')
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.defaultName', dynamicWebDefaultDocument)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.documentSource',dynamicWebDocumentSource)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.sourceViewable', dynamicWebSourceViewable)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.executionController', new PhpExecutionController()) // Adds PHP predefined variables
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.clientCachingMode', dynamicWebClientCachingMode)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.fileUploadSizeThreshold', fileUploadSizeThreshold)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.handlersDocumentSource', handlersDocumentSource)
applicationGlobals.put('com.threecrickets.prudence.GeneratedTextResource.cacheKeyPatternHandlers', cacheKeyPatternHandlers)

var dynamicWeb = new Finder(applicationInstance.context, classLoader.loadClass('com.threecrickets.prudence.GeneratedTextResource'))
dynamicWebBaseURL = fixURL(dynamicWebBaseURL)
router.attachBase(dynamicWebBaseURL, dynamicWeb)

if(dynamicWebDefrost) {
	var defrostTasks = DefrostTask.forDocumentSource(dynamicWebDocumentSource, languageManager, 'javascript', true, true)
	for(var i in defrostTasks) {
		tasks.push(defrostTasks[i])
	}
}

//
// Static web
//

var staticWeb = new Directory(applicationInstance.context, new File(applicationBasePath + staticWebBasePath).toURI().toString())
staticWeb.listingAllowed = staticWebDirectoryListingAllowed
staticWeb.negotiatingContent = true
staticWebBaseURL = fixURL(staticWebBaseURL)
if(staticWebCompress) {
	router.filterBase(staticWebBaseURL, new Encoder(null), staticWeb)
}
else {
	router.attachBase(staticWebBaseURL, staticWeb)
} 

//
// Resources
//

var resourcesDocumentSource = new DocumentFileSource(applicationBase + resourcesBasePath, applicationBasePath + resourcesBasePath, resourcesDefaultName, 'js', resourcesMinimumTimeBetweenValidityChecks)
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.languageManager', languageManager)
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.defaultLanguageTag', 'javascript')
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.defaultName', resourcesDefaultName)
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.documentSource', resourcesDocumentSource)
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.sourceViewable', resourcesSourceViewable)
applicationGlobals.put('com.threecrickets.prudence.DelegatedResource.fileUploadSizeThreshold', fileUploadSizeThreshold)

resources = new Finder(applicationInstance.context, classLoader.loadClass('com.threecrickets.prudence.DelegatedResource'))
resourcesBaseURL = fixURL(resourcesBaseURL)
router.attachBase(resourcesBaseURL, resources)

if(resourcesDefrost) {
	var defrostTasks = DefrostTask.forDocumentSource(resourcesDocumentSource, languageManager, 'javascript', false, true)
	for(var i in defrostTasks) {
		tasks.push(defrostTasks[i])
	}
}

//
// SourceCode
//

if(showDebugOnError) {
	var documentSources = new ArrayList()
	documentSources.add(dynamicWebDocumentSource)
	documentSources.add(resourcesDocumentSource)
	applicationGlobals.put('com.threecrickets.prudence.SourceCodeResource.documentSources', documentSources)
	var sourceCode = new Finder(applicationInstance.context, classLoader.loadClass('com.threecrickets.prudence.SourceCodeResource'))
	showSourceCodeURL = fixURL(showSourceCodeURL)
	router.attach(showSourceCodeURL, sourceCode).matchingMode = Template.MODE_EQUALS
}

//
// Preheat
//

if(dynamicWebPreheat) {
	var preheatTasks = PreheatTask.forDocumentSource(dynamicWebDocumentSource, component.context, applicationInternalName)
	for(var i in preheatTasks) {
		tasks.push(preheatTasks[i])
	}
}

for(var i in preheatResources) {
	var preheatResource = preheatResources[i]
	tasks.push(new PreheatTask(component.context, applicationInternalName, preheatResource))
}

File: python/defaults/application/routing.py
#
# Prudence Application Routing
#

from java.lang import ClassLoader
from java.io import File
from javax.script import ScriptEngineManager

from org.restlet.routing import Router, Redirector, Template
from org.restlet.resource import Finder, Directory
from com.threecrickets.scripturian import DefrostTask
from com.threecrickets.scripturian.file import DocumentFileSource
from com.threecrickets.prudence.util import PrudenceRouter, PreheatTask

classLoader = ClassLoader.getSystemClassLoader()

#
# Utilities
#

# Makes sure we have slashes where we expect them
def fix_url(url):
	url = url.replace('//', '/') # no doubles
	if len(url) > 0 and url[0] == '/': # never at the beginning
		url = url[1:]
	if len(url) > 0 and url[-1] != '/': # always at the end
		url = url + '/'
	return url

#
# Internal router
#

component.internalRouter.attach('/' + application_internal_name, application + '/').matchingMode = Template.MODE_STARTS_WITH

#
# Hosts
#
# Note that the application's context will not be created until we attach the application to at least one
# virtual host. See component/hosts.py for more information.
#

add_trailing_slash = Redirector(application.context, '{ri}/', Redirector.MODE_CLIENT_PERMANENT)

sys.stdout.write('%s: ' % application.name)
for i in range(len(hosts)):
	host, url = hosts.items()[i]
	if url is None:
		url = application_default_url
	sys.stdout.write('"%s" on %s' % (url, host.name))
	host.attach(url, application).matchingMode = Template.MODE_STARTS_WITH
	if url != '/':
		if url[-1] == '/':
			url = url[:-1]
		host.attach(url, add_trailing_slash).matchingMode = Template.MODE_EQUALS
	if i < len(hosts) - 1:
		sys.stdout.write(', ')
print '.'

attributes = application.context.attributes

#
# Inbound root
#

router = PrudenceRouter(application.context)
router.routingMode = Router.MODE_BEST_MATCH
application.inboundRoot = router

#
# Add trailing slashes
#

for url in url_add_trailing_slash:
	url = fix_url(url)
	if len(url) > 0:
		if url[-1] == '/':
			url = url[:-1]
		router.attach(url, add_trailing_slash)

#
# Dynamic web
#

script_engine_manager = ScriptEngineManager()
dynamic_web_document_source = DocumentFileSource(application_base_path + dynamic_web_base_path, dynamic_web_default_document, dynamic_web_minimum_time_between_validity_checks)
attributes['com.threecrickets.prudence.GeneratedTextResource.engineManager'] = script_engine_manager
attributes['com.threecrickets.prudence.GeneratedTextResource.defaultEngineName'] = 'python'
attributes['com.threecrickets.prudence.GeneratedTextResource.defaultName'] = dynamic_web_default_document
attributes['com.threecrickets.prudence.GeneratedTextResource.documentSource'] = dynamic_web_document_source
attributes['com.threecrickets.prudence.GeneratedTextResource.sourceViewable'] = dynamic_web_source_viewable

dynamic_web = Finder(application.context, classLoader.loadClass('com.threecrickets.prudence.GeneratedTextResource'))
router.attach_base(fix_url(dynamic_web_base_url), dynamic_web)

if dynamic_web_defrost:
	for defrost_task in DefrostTask.create(dynamic_web_document_source, script_engine_manager, True):
		tasks.push(defrost_task)

#
# Static web
#

static_web = Directory(application.context, File(application_base_path + static_web_base_path).toURI().toString())
static_web.listingAllowed = static_web_directory_listing_allowed
static_web.negotiateContent = True
router.attach_base(fix_url(static_web_base_url), static_web)

#
# Resources
#

resources_document_source = DocumentFileSource(application_base_path + resources_base_path, resources_default_name, resources_minimum_time_between_validity_checks)
attributes['com.threecrickets.prudence.DelegatedResource.engineManager'] = script_engine_manager
attributes['com.threecrickets.prudence.DelegatedResource.defaultEngineName'] = 'python'
attributes['com.threecrickets.prudence.DelegatedResource.defaultName'] = resources_default_name
attributes['com.threecrickets.prudence.DelegatedResource.documentSource'] = resources_document_source
attributes['com.threecrickets.prudence.DelegatedResource.sourceViewable'] = resources_source_viewable

resources = Finder(application.context, classLoader.loadClass('com.threecrickets.prudence.DelegatedResource'))
router.attach_base(fix_url(resources_base_url), resources)

if resources_defrost:
	for defrost_task in DefrostTask.create(resources_document_source, script_engine_manager, True):
		tasks.push(defrost_task)

#
# Preheat
#

if dynamic_web_preheat:
	for preheat_task in PreheatTask.create(component.context, application_internal_name, dynamic_web_document_source):
		tasks.push(preheat_task)

for preheat_resource in preheat_resources:
	tasks.append(PreheatTask(component.context, application_internal_name, preheat_resource))
